src/app/models/form.server.model.ts

```diff
@@ -586,6 +586,37 @@ const compileFormModel = (db: Mongoose): IFormModel => {
       return next(err)
     }
 
+    // Check if MyInfo fields exist and if they do, the total number of fields <= 30, and that
+    // the response mode and authType fields are set appropriately.
+    let myInfoFieldCount = 0
+    if (this.form_fields !== undefined) {
+      this.form_fields.forEach((field) => {
+        if (field.myInfo !== undefined) {
+          myInfoFieldCount++
+        }
+      })
+    }
+
+    if (
+      myInfoFieldCount > 0 &&
+      (this.authType !== AuthType.MyInfo ||
+        this.responseMode !== ResponseMode.Encrypt)
+    ) {
+      const validationError = this.invalidate(
+        'form_fields',
+        'This form contains MyInfo fields but does not have MyInfo enabled or is a storage mode form.',
+      ) as mongoose.Error.ValidationError
+      return next(validationError)
+    }
+
+    if (myInfoFieldCount > 30) {
+      const validationError = this.invalidate(
+        'form_fields',
+        'This form contains too many MyInfo fields. A maximum of 30 MyInfo fields are allowed.',
+      ) as mongoose.Error.ValidationError
+      return next(validationError)
```

Comment:
actually should we change this to be a key-specific validator for `form_fields` instead? I think this pre-validate hook looks at risk of becoming a mash of unrelated functionality. you can look at this chunk of code in `src/app/models/field/tableField.ts` for an example of how to access `this` in a key-specific validator:
```
maximumRows: {
      type: Number,
      min: 2,
      validate: {
        validator: function (this: ITableFieldSchema, v?: number) {
          return !v || v > this.minimumRows
        },
        message: 'Maximum number of rows must be greater than minimum.',
      },
    },
```

Reply:
Currently `form_fields` is defined as `form_fields: [BaseFieldSchema]` and I am not sure you can add a validator there easily?

So the idea is to move this code into `BaseFieldSchema`, but instead of sticking it into https://github.com/opengovsg/FormSG/blob/32a18fc781a125046d343ede2fed03205533be42/src/app/models/field/baseField.ts#L73-L88, you want to create another property of BaseFieldSchema and add the validator for the count there? 

Isn't that more complicated than either adding the validator to `MyInfoSchema` (what was originally written) or leaving it here?

Reply:
the current code is this:
```
form_fields: [BaseFieldSchema],
```

since we're performing validation on form fields, I'm thinking something we can change it to something like this:
```
form_fields: {
        type: [BaseFieldSchema],
        validate: {
          validator: function (this: IFormSchema) {
            const myInfoFieldCount = (this.form_fields ?? []).reduce(
              (acc, field) => acc + (field.myInfo ? 1 : 0),
              0,
            )
            if (myInfoFieldCount === 0) return true
            return (
              this.authType === AuthType.MyInfo &&
              this.responseMode === ResponseMode.Email &&
              myInfoFieldCount <= 30
            )
          },
          message:
            'Check that your form is Myinfo-authenticated, is an email mode form and has 30 or fewer MyInfo fields.',
        },
      },
```

what do you think? the idea is that since the validation is specific to `form_fields`, we should add it to the validator for that particular key.

Reply:
Yeah I guess this is fair. Let me make the change.